packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,376 @@
+import * as Server from '@ucanto/server'
+import { ed25519 } from '@ucanto/principal'
+import * as Blob from '@web3-storage/capabilities/blob'
+import * as W3sBlob from '@web3-storage/capabilities/web3.storage/blob'
+import * as HTTP from '@web3-storage/capabilities/http'
+import * as API from '../types.js'
+
+import { createConcludeInvocation } from '../ucan/conclude.js'
+import { BlobSizeOutsideOfSupportedRange, AwaitError } from './lib.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAdd, API.BlobAddSuccess, API.BlobAddFailure>}
+ */
+export function blobAddProvider(context) {
+  return Server.provideAdvanced({
+    capability: Blob.add,
+    handler: async ({ capability, invocation }) => {
+      // Prepare context
+      const { blob } = capability.nb
+      const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+        Server.DID.parse(capability.with).did()
+      )
+
+      // Verify blob is within accept size
+      if (blob.size > context.maxUploadSize) {
+        return {
+          error: new BlobSizeOutsideOfSupportedRange(context.maxUploadSize),
+        }
+      }
+
+      // Create next tasks
+      const next = await createNextTasks({
+        context,
+        blob,
+        space,
+        cause: invocation.link(),
+      })
+
+      // Schedule allocate
+      const scheduleAllocateRes = await scheduleAllocate({
+        context,
+        blob,
+        allocate: next.allocate,
+        allocatefx: next.allocatefx,
+      })
+      if (scheduleAllocateRes.error) {
+        return scheduleAllocateRes
+      }
```

Comment:
Nit: I think it would be easier to follow what's going on if instead you refactored code into following three functions

```ts
const accept = (context, { blob, space, cause }): { task: Invocation, receipt?: Receipt }
```

And same for the `put` and `allocate`. Specifically I'm suggesting to combine invocation creation and with what `schedule` functions do and return back invocation and receipt when possible.